docs(plugins): update plugin documentation, api reference, and developer guide (#542)
falcosecurity/falco-website: content/en/docs/plugins/plugin-api-reference.md

##### `const char* plugin_get_fields() [Required: no]`

> This function should return the set of fields supported by the plugin. Remember, a field is a name (e.g. `proc.name`) that can extract a value (e.g. `nginx`) from an event (e.g. a syscall event involving a process). The format is a json string, with memory owned by the plugin, which contains an array of objects. Each object describes one field. Here's an example:

Reviewer: ```suggestion
This function should return the set of fields supported by the plugin. Remember, a field is a name (e.g. `proc.name`) that can extract a value (e.g. `nginx`) from an event (e.g. a syscall event involving a process). The return value is a string whose memory is owned by the plugin. The string is json formatted and contains an array of objects. Each object describes one field. Here's an example:
```

Reviewer: Not sure my suggestion is good. Just a personal preference. Feel free to ignore my suggestion :)

* On failure, make sure to return a meaningful error message in the next call to `plugin_get_last_error()`.
* On failure, the plugin framework will *not* call plugin_destroy, so do not return any allocated state.

> The format of the config string is entirely determined by the plugin author, and by default is passed unchanged from Falco/the application using the plugin framework to the plugin. However, semi-structured formats like JSON/YAML are preferable to free-form text. In those cases, the plugin author can provide a schema describing the config string contents by implementing the optional `get_init_schema` function. If so, the `init()` function can assume the passed-in configuration string to always be well-formed, and can avoid performing any error handling. The plugin framework will take care of automatically parsing it against the provided schema and generating ad-hoc errors accordingly. Please refer to the documentation of `get_init_schema` for more details.

Reviewer: Shall we recommend JSON?

Author: I think it is pretty much recommended right now. By using JSON, you have access to the schema feature using JSON Schema, which delegates the configuration validation to the framework. I think this the `get_init_schema` doc section make this explicit, but if you want I can add few lines here too.

* `datalen`: pointer to a 32bit integer. The plugin will set it the size of the buffer pointed by data.
* `ts`: the event timestamp, in nanoseconds since the epoch. Can be (uint64_t)-1, in which case the engine will automatically fill the event time with the current time.

> It is not necessary to fill in the evtnum struct member when returning events via plugin_next_batch. Event numbers are assigned by the plugin framework.

Reviewer: Is that just unnecessary? Or the plugin author must NOT do that?

Author: More that unnecessary, I would say useless. The value would be overwritten even if the plugin writes into it.
